google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/PubsubMessageWrapper.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub.v1;
+
+import com.google.api.core.InternalApi;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.SubscriptionName;
+import com.google.pubsub.v1.TopicName;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+
+/**
+ * A wrapper class for a {@link PubsubMessage} object that handles creation and tracking of
+ * OpenTelemetry {@link Span} objects for different operations that occur during publishing.
+ */
+public class PubsubMessageWrapper {
+  private PubsubMessage message;
+
+  private final TopicName topicName;
+  private final SubscriptionName subscriptionName;
+
+  // Attributes set only for messages received from a streaming pull response.
+  private final String ackId;
+  private final int deliveryAttempt;
+
+  private static final String PUBLISH_START_EVENT = "publish start";
+  private static final String PUBLISH_END_EVENT = "publish end";
+
+  private static final String MODACK_START_EVENT = "modack start";
+  private static final String MODACK_END_EVENT = "modack end";
+  private static final String NACK_START_EVENT = "nack start";
+  private static final String NACK_END_EVENT = "nack end";
+  private static final String ACK_START_EVENT = "ack start";
+  private static final String ACK_END_EVENT = "ack end";
+
+  private static final String GOOGCLIENT_PREFIX = "googclient_";
+
+  private static final String MESSAGE_RESULT_ATTR_KEY = "messaging.gcp_pubsub.result";
+
+  private Span publisherSpan;
+  private Span publishFlowControlSpan;
+  private Span publishBatchingSpan;
+
+  private Span subscriberSpan;
+  private Span subscribeConcurrencyControlSpan;
+  private Span subscribeSchedulerSpan;
+  private Span subscribeProcessSpan;
+
+  @InternalApi("For use by the google-cloud-pubsub library only")
+  public PubsubMessageWrapper(Builder builder) {
```

Comment:
This class and all the methods can be package private if it is intended to be used by pubsub only. Same thing for other internal classes.

Reply:
Did this for `PubsubMessageWrapper` and `OpenTelemetryPubsubTracer`. We may be able to do this for some other internal class, but I think that is out of scope for this PR.